### PR TITLE
FINERACT-1960: GL account mappings to use Savings product with Accrua…

### PR DESCRIPTION
**Pull request by alberto-art3ch**
…l accounting

## Description

Financial institutions will want to recognize expenses/incomes accrued on deposit product accounts too similar to the Loan products.

Today, In Fineract deposit products do not support accrual accounting. This spec is to enhance the accrual accounting in Fineract to support periodic accrual of expenses/incomes on deposit accounts too.

Deposit Products refers to the Savings products, Fixed Deposit Products, and Recurring Deposit Products of Fineract Application, and accrual accounting option should be available for all three types.

Deposit product expenses: Interest to be paid on deposit accounts

Deposit product incomes: fees, interest on overdrafts

[FINERACT-1960](https://issues.apache.org/jira/browse/FINERACT-1960)

Checklist
[+] Include the three new parameters in the Deposit product creation and edition
[+] Add the support for Accrual accounting in the Deposit product and GL account mappings

Screenshot 2024-02-17 at 8 59 39 p m

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by bharathcgowda** [comment]
@adamsaghy  could you please in reviewing this PR when ever you are available?

**Comment by alberto-art3ch** [comment]
> @adamsaghy could you please in reviewing this PR when ever you are available?

@adamsaghy PR reviewed and updated, Thanks!

File: fineract-accounting/src/main/java/org/apache/fineract/accounting/producttoaccountmapping/service/ProductToGLAccountMappingReadPlatformServiceImpl.java [hunk 1]
```diff
@@ -225,49 +229,44 @@ public Map<String, Object> fetchAccountMappingDetailsForLoanProduct(final Long l
...
-        final Map<String, Object> accountMappingDetails = new LinkedHashMap<>(8);
-
         final ProductToGLAccountMappingMapper rm = new ProductToGLAccountMappingMapper();
         final String sql = "select " + rm.schema() + " and product_id = ? and payment_type is null and mapping.charge_id is null ";
 
         final List<Map<String, Object>> listOfProductToGLAccountMaps = this.jdbcTemplate.query(sql, rm, // NOSONAR
                 new Object[] { PortfolioProductType.SAVING.getValue(), savingsProductId });
 
-        if (AccountingRuleType.CASH_BASED.getValue().equals(accountingType)) {
+        Map<String, Object> accountMappingDetails = setBaseSavingsProductToGLAccountMaps(listOfProductToGLAccountMaps);
```
Comment (adamsaghy) [inline fineract-accounting/src/main/java/org/apache/fineract/accounting/producttoaccountmapping/service/ProductToGLAccountMappingReadPlatformServiceImpl.java]: Please do not reuse cash based accounting settings. 

Define the mapping for Cash based and for Accrual based separately. Even if there are common rules and accounts, still it is better to handle them and define them differently!

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java [hunk 2]
```diff
@@ -573,6 +574,27 @@ public void createCashBasedJournalEntriesAndReversalsForSavingsTax(final Office
...
 
+    public void createAccrualBasedJournalEntriesAndReversalsForSavingsTax(final Office office, final String currencyCode,
+            final AccrualAccountsForSavings accountTypeToBeDebited, final AccrualAccountsForSavings accountTypeToBeCredited,
+            final Long savingsProductId, final Long paymentTypeId, final Long savingsId, final String transactionId,
+            final LocalDate transactionDate, final BigDecimal amount, final Boolean isReversal, final List<TaxPaymentDTO> taxDetails) {
+
+        for (TaxPaymentDTO taxPaymentDTO : taxDetails) {
+            if (taxPaymentDTO.getAmount() != null) {
+                if (taxPaymentDTO.getCreditAccountId() == null) {
+                    createCashBasedCreditJournalEntriesAndReversalsForSavings(office, currencyCode, accountTypeToBeCredited.getValue(),
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java]: An accrual based journal entry should not call cashbased once... Please extract the logic into a common place if needed, but Cash based accounting and Accrual based accounting should not be mixed together!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/FixedDepositProductsApiResourceSwagger.java [hunk 3]
```diff
@@ -318,88 +318,31 @@ private GetFixedDepositProductsProductIdInterestCompoundingPeriodType() {}
...
-                public Long id;
-                @Schema(example = "savings ref tool kit")
-                public String name;
-                @Schema(example = "21")
-                public String glCode;
-            }
-
-            static final class GetFixedDepositProductsProductIdTransfersInSuspenseAccount {
+            @Schema(example = "12")
+            public Integer id;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/FixedDepositProductsApiResourceSwagger.java]: GL account id is long!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/RecurringDepositProductsApiResourceSwagger.java [hunk 4]
```diff
@@ -343,88 +343,30 @@ private GetRecurringDepositProductsProductIdInterestCompoundingPeriodType() {}
...
-                public Long id;
-                @Schema(example = "savings ref tool kit")
-                public String name;
-                @Schema(example = "21")
-                public String glCode;
-            }
-
-            static final class GetRecurringDepositProductsProductIdTransfersInSuspenseAccount {
+            @Schema(example = "12")
+            public Integer id;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/RecurringDepositProductsApiResourceSwagger.java]: Id is Long!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/DepositProductDataValidator.java [hunk 5]
```diff
@@ -115,7 +113,7 @@ public void validateForFixedDepositCreate(final String json) {
 
         final JsonElement element = this.fromApiJsonHelper.parse(json);
 
-        validateDepositDetailForCreate(element, this.fromApiJsonHelper, baseDataValidator);
+        validateDepositDetailForCreate(element, this.fromApiJsonHelper, baseDataValidator, DepositAccountType.RECURRING_DEPOSIT);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/DepositProductDataValidator.java]: Validation of Fixed deposit is the method name but here you are providing Recurring deposit. Why?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/DepositAccountReadPlatformServiceImpl.java [hunk 6]
```diff
@@ -479,6 +480,7 @@ public Collection<AccountTransferDTO> retrieveDataForInterestTransfer() {
         sqlBuilder.append(
                 "and st.transaction_type_enum = ? and sa.status_enum = ? and st.is_reversed=false and st.transaction_date > coalesce(sa.lockedin_until_date_derived,sa.activatedon_date)");
 
+        log.info("SQL {}", sqlBuilder.toString());
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/DepositAccountReadPlatformServiceImpl.java]: Why to log at info level?

**Review by adamsaghy — COMMENTED** [review]
Please see my comments!

Also there are a lot of changes, so my confidence in my review is not 100%!

It would be nice to see a summary of changes!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/DepositAccountReadPlatformServiceImpl.java (hunk 6 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/DepositAccountReadPlatformServiceImpl.java]: Done!

File: fineract-accounting/src/main/java/org/apache/fineract/accounting/producttoaccountmapping/service/ProductToGLAccountMappingReadPlatformServiceImpl.java (hunk 1 above)
Comment (alberto-art3ch) [reply fineract-accounting/src/main/java/org/apache/fineract/accounting/producttoaccountmapping/service/ProductToGLAccountMappingReadPlatformServiceImpl.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java (hunk 2 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/FixedDepositProductsApiResourceSwagger.java (hunk 3 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/FixedDepositProductsApiResourceSwagger.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/RecurringDepositProductsApiResourceSwagger.java (hunk 4 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/RecurringDepositProductsApiResourceSwagger.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/DepositProductDataValidator.java (hunk 5 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/DepositProductDataValidator.java]: Done!

File: fineract-accounting/src/main/java/org/apache/fineract/accounting/producttoaccountmapping/service/ProductToGLAccountMappingReadPlatformServiceImpl.java (hunk 1 above)
Comment (alberto-art3ch) [reply fineract-accounting/src/main/java/org/apache/fineract/accounting/producttoaccountmapping/service/ProductToGLAccountMappingReadPlatformServiceImpl.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java (hunk 2 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/FixedDepositProductsApiResourceSwagger.java (hunk 3 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/FixedDepositProductsApiResourceSwagger.java]: Done!

**Review by alberto-art3ch — COMMENTED** [review]
Done! Comments attended, Thanks

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**